src/EditorFeatures/Text/ITextImageHelpers.cs

```diff
@@ -51,35 +52,54 @@ public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion o
         }
 
         if (changes == null)
-        {
             return [];
-        }
-        else
+
+        var builder = ArrayBuilder<TextChangeRange>.GetInstance(changes.Count);
+        for (var i = 0; i < changes.Count; i++)
         {
-            return ImmutableArray.CreateRange(changes.Select(forward ? s_forwardTextChangeRange : s_backwardTextChangeRange));
+            var change = changes[i];
+            builder.Add(forward ? s_forwardTextChangeRange(change) : s_backwardTextChangeRange(change));
         }
+
+        return builder.ToImmutableAndFree();
     }
 
     private static TextChangeRange GetChangeRanges(ITextImageVersion oldVersion, ITextImageVersion newVersion, bool forward)
     {
         TextChangeRange? range = null;
-        var iterator = GetMultipleVersionTextChanges(oldVersion, newVersion, forward);
-        foreach (var changes in forward ? iterator : iterator.Reverse())
+        using var _ = ArrayBuilder<ArrayBuilder<TextChangeRange>>.GetInstance(out var builder);
+
+        GetMultipleVersionTextChanges(oldVersion, newVersion, forward, builder);
+        foreach (var changes in builder)
         {
             range = range.Accumulate(changes);
+            changes.Free();
```

Comment:
nit.  consider moving into a try/finally.

Reply:
Unless you feel strongly, I'd rather keep as. Exceptions shouldn't be happening during the Accumulate call, and it they do, not putting a couple arrays back in the pool is probably ok.

Reply:
i do not feel strongly here.  this is simple enough :)